packages/react-router/modules/Route.js

```diff
@@ -92,20 +93,24 @@ class Route extends React.Component {
     );
   }
 
-  componentWillReceiveProps(nextProps, nextContext) {
+  componentDidUpdate(prevProps) {
     warning(
-      !(nextProps.location && !this.props.location),
+      !(prevProps.location && !this.props.location),
       '<Route> elements should not change from uncontrolled to controlled (or vice versa). You initially used no "location" prop and then provided one on a subsequent render.'
     );
 
     warning(
-      !(!nextProps.location && this.props.location),
+      !(!prevProps.location && this.props.location),
       '<Route> elements should not change from controlled to uncontrolled (or vice versa). You provided a "location" prop initially but omitted it on a subsequent render.'
     );
 
-    this.setState({
-      match: this.computeMatch(nextProps, nextContext.router)
-    });
+    const newMatch = this.computeMatch(this.props, this.context.router);
```

Comment:
I don't really like re-computing the `match` in the commit phase; this would spend a render phase rendering based on the outdated `state.match` just to throw away that work and re-render using the real match.

If it wasn't for the `getChildContext()` to support legacy `context`, I would probably just compute the `match` in `render()` (maybe looking into memoizing computing matches). This is probably a decent use case for `getDerivedStateFromProps()`, but would break apps apps that use pre-`16.3` version of React (or is there a way to backport `gDSFP`?).

Reply:
There's `react-lifecycles-compat`, but it's rather large.

We've found most folks are on 16.3+ over on React Redux, FWIW.

Reply:
Yeah, to my understanding, this is probably the best way to do things without a breaking change (and without including lifecycles-compat).

I'm open to ideas though!